File: day-03/day-03.py
import re

with open('data.txt', 'r') as r:
    lines = r.read()
    matches = re.findall(r"mul\((\d+),(\d+)\)", lines)
    part1 = sum(int(x) * int(y) for x, y in matches)
    print("Part 1:", part1)

    instructions = re.findall(r"(do\(\)|don't\(\)|mul\((\d+),(\d+)\))", lines)
    mul_enabled = True
    part2 = 0
    for instruction in instructions:
        if instruction[0] == "do()":
            mul_enabled = True
        elif instruction[0] == "don't()":
            mul_enabled = False
        elif "mul" in instruction[0] and mul_enabled:
            x, y = map(int, instruction[1:])
            part2 += x * y
    print("Part 2:", part2)

File: day-03/day-03.rs
use std::fs;
use regex::Regex;

fn main() {
    let input: String = fs::read_to_string("day-03/data.txt").unwrap();
    let lines: Vec<&str> = input.lines().collect::<Vec<&str>>();
    let line: String = lines.join("");
    let mut part1: i32 = 0;
    let mut part2: i32 = 0;

    let re1: Regex = Regex::new(r"mul\((\d+),(\d+)\)").unwrap();
    
    re1.captures_iter(&line).for_each(|x| {
        let a: i32 = x[1].parse::<i32>().unwrap();
        let b: i32 = x[2].parse::<i32>().unwrap();
        part1 += a * b;
    });

    let re2: Regex = Regex::new(r"(do\(\)|don't\(\)|mul\((\d+),(\d+)\))").unwrap();
    let mut enabled: bool = true;

    re2.captures_iter(&line).for_each(|x| {
        let instruction: &str = &x[0];
        if instruction.contains("do()") {
            enabled = true;
        }
        else if instruction.contains("don't()") {
            enabled = false;
        }
        else {
            let a: i32 = x[2].parse::<i32>().unwrap();
            let b: i32 = x[3].parse::<i32>().unwrap();
            if enabled {
                part2 += a * b;
            }
        }
    });

    println!("Part 1: {}", part1);
    println!("Part 2: {}", part2);

}